engine/api/download/download.go

```diff
@@ -0,0 +1,119 @@
+
+	"github.com/ovh/cds/sdk"
+)
+
+type Conf struct {
+	Directory             string
+	DownloadFromGitHub    bool
+	ArtifactoryURL        string
+	ArtifactoryPath       string
+	ArtifactoryRepository string
+	ArtifactoryToken      string
+	SupportedOSArch       string
+}
+
+func Init(ctx context.Context, conf Conf) error {
+	// Checking downloadable binaries
+	sdk.InitSupportedOSArch(conf.SupportedOSArch)
+	if err := CheckBinary(ctx, conf, "worker", sdk.GOOS, sdk.GOARCH, ""); err != nil {
+		return err
+	}
+	resources := sdk.AllDownloadableResourcesWithAvailability(conf.Directory)
+	var hasWorker, hasCtl, hasEngine bool
+	for _, r := range resources {
+		if r.Available != nil && *r.Available {
+			switch r.Name {
+			case "worker":
+				hasWorker = true
+			case "cdsctl":
+				hasCtl = true
+			case "engine":
+				hasEngine = true
+			}
+		}
+	}
+	if !hasEngine {
+		log.Error(ctx, "engine is unavailable for download, this may lead to a poor user experience. Please check your configuration file or the %s directory", conf.Directory)
+	}
+	if !hasCtl {
+		log.Error(ctx, "cdsctl is unavailable for download, this may lead to a poor user experience. Please check your configuration file or the %s directory", conf.Directory)
+	}
+	if !hasWorker {
+		// If no worker, let's exit because CDS for run anything
+		log.Error(ctx, "worker is unavailable for download. Please check your configuration file or the %s directory", conf.Directory)
+		return errors.New("worker binary unavailable")
+	}
+	return nil
+}
+
+// CheckBinary checks if binary exist and download it if needed
+func CheckBinary(ctx context.Context, conf Conf, name, os, arch, variant string) error {
+	filename := sdk.BinaryFilename(name, os, arch, variant)
+	if sdk.IsDownloadedBinary(conf.Directory, filename) {
+		return nil
+	}
+
+	var filenameToDownload string
+	if name == "worker" {
+		filenameToDownload = "cds-worker-all.tar.gz"
+	} else {
+		filenameToDownload = filename
+	}
+
+	if conf.DownloadFromGitHub {
+		log.Info(ctx, "downloading %v from GitHub into %v", filenameToDownload, conf.Directory)
+		if err := sdk.DownloadFromGitHub(ctx, conf.Directory, filenameToDownload, sdk.VERSION); err != nil {
+			return err
+		}
+	} else if conf.ArtifactoryURL != "" {
+		log.Info(ctx, "downloading %v from Artifactory into %v", filenameToDownload, conf.Directory)
+		if err := GetBinaryFromArtifactory(conf, filenameToDownload); err != nil {
+			return err
+		}
+	}
+
+	if strings.HasSuffix(filenameToDownload, ".tar.gz") {
+		log.Info(ctx, "untar %v", filenameToDownload)
+		if err := archiver.DefaultTarGz.Unarchive(path.Join(conf.Directory, filenameToDownload), conf.Directory); err != nil {
+			return sdk.WrapError(err, "unarchive %s failed", filenameToDownload)
+		}
+	}
+
+	return nil
+}
+
+func GetBinaryFromArtifactory(conf Conf, filename string) error {
+	jlog.SetLogger(jlog.NewLogger(jlog.ERROR, os.Stdout))
+	artiClient, err := art.CreateArtifactoryClient(conf.ArtifactoryURL, conf.ArtifactoryToken)
```

Comment:
Do not use contrib package in API.
Maybe refact/reuse code in package api/integration/artifact_manager package